x-pack/platform/plugins/shared/onechat/public/application/components/tools/form/registry/tools_form_registry.tsx

```diff
@@ -0,0 +1,71 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { ToolType } from '@kbn/onechat-common';
+
+import type { CreateToolPayload, UpdateToolPayload } from '../../../../../../common/http_api/tools';
+import type { ToolFormData } from '../types/tool_form_types';
+
+import { esqlToolFormRegistryEntry } from './tool_types/esql';
+import type { SupportedToolTypeRegistryEntry } from './common';
+import { indexSearchToolRegistryEntry } from './tool_types/index_search';
+import { builtinToolRegistryEntry } from './tool_types/builtin';
+
+export const TOOLS_FORM_REGISTRY: Record<ToolType, SupportedToolTypeRegistryEntry> = {
+  [ToolType.esql]: esqlToolFormRegistryEntry,
+  [ToolType.index_search]: indexSearchToolRegistryEntry,
+  // This is only for displaying read-only built-in tools
+  [ToolType.builtin]: builtinToolRegistryEntry,
+};
+
+export function getToolTypeConfig(toolType: ToolType) {
+  return TOOLS_FORM_REGISTRY[toolType];
+}
+
+export function getCreatePayloadFromData(
+  data: ToolFormData,
+  toolType?: ToolType
+): CreateToolPayload {
+  const type = toolType || data.type;
+  const config = getToolTypeConfig(type);
+
+  if (!config) {
+    throw new Error(`Unknown tool type: ${type}`);
+  }
+  // @ts-expect-error TS2345 - Union type ToolFormData cannot be narrowed to specific resolver type at compile time
```

Comment:
i was super lost with that one ... maybe there is a more robust pattern to bypass this error ?